gr-fec/lib/ldpc_H_matrix_impl.cc

```diff
@@ -182,38 +179,39 @@ void ldpc_H_matrix_impl::back_solve_mod2(gsl_matrix* x,
     // vector. Solve for x.
 
     // Allocate memory for the result
-    int num_rows = (*U).size1;
-    int num_cols_U = (*U).size2;
+    const int num_rows = (*U).size1;
+    const int num_cols_U = (*U).size2;
 
     // Back solve
     for (int i = num_rows - 1; i >= 0; i--) {
         // x[i] = y[i]
-        gsl_matrix_set(x, i, 0, gsl_matrix_get(y, i, 0));
-
-        int j;
-        for (j = i + 1; j < num_cols_U; j++) {
-            int U_i_j = gsl_matrix_get(U, i, j);
-            int x_i = gsl_matrix_get(x, i, 0);
-            int x_j = gsl_matrix_get(x, j, 0);
-            int temp1 = (U_i_j * x_j) % 2;
-            int temp2 = (x_i + temp1) % 2;
-            gsl_matrix_set(x, i, 0, temp2);
+        gsl_matrix_uchar_set(x, i, 0, gsl_matrix_uchar_get(y, i, 0));
+
+        for (int j = i + 1; j < num_cols_U; j++) {
+            const int U_i_j = gsl_matrix_uchar_get(U, i, j);
+            const int x_i = gsl_matrix_uchar_get(x, i, 0);
+            const int x_j = gsl_matrix_uchar_get(x, j, 0);
+            const int temp1 = (U_i_j * x_j) & 1;
+            const int temp2 = (x_i + temp1) & 1;
+            gsl_matrix_uchar_set(x, i, 0, temp2);
         }
         // Perform x[i] /= U[i,i], GF(2) operations
-        int U_i_i = gsl_matrix_get(U, i, i);
-        int x_i = gsl_matrix_get(x, i, 0);
-        if (x_i == 0 && U_i_i == 1)
-            gsl_matrix_set(x, i, 0, 0);
-        else if (x_i == 0 && U_i_i == 0)
-            gsl_matrix_set(x, i, 0, 0);
-        else if (x_i == 1 && U_i_i == 1)
-            gsl_matrix_set(x, i, 0, 1);
-        else if (x_i == 1 && U_i_i == 0)
-            std::cout << "Error in "
-                      << " ldpc_H_matrix_impl::back_solve_mod2,"
-                      << " division not defined.\n";
-        else
-            std::cout << "Error in ldpc_H_matrix::back_solve_mod2\n";
+        int U_i_i = gsl_matrix_uchar_get(U, i, i);
+        int x_i = gsl_matrix_uchar_get(x, i, 0);
+        // 0 divided by anything is 0 in GF2, so, hm, only care about 1
+        if (x_i == 1) {
+            if (U_i_i == 1) {
+                gsl_matrix_uchar_set(x, i, 0, 1);
+            } else {
+                // x[i] = 1 and U[i][i] = 0 -> a/0 = undef
+                gr::logger("ldpc_H_matrix_impl")
+                    .error("Error in back_solve_mod2, division between x[{0}]={1} and "
+                           "U[{0}]={2} not defined.",
+                           i,
+                           x_i,
+                           U_i_i);
+            }
+        }
```

Comment:
To me it seems that this won't catch `0/0`. I would do it as follows:

1. If `U_i_i == 0`, then either log an error or throw an exception. I would need to think more carefully about in which context this function is being run and whether failure to divide should be a fatal error (for instance, can it happen normally as a decoding failure with a codeword with many bit errors, or mathematically it should never happen and if we see it then something has gone deeply wrong?).

2. Do nothing. Technically we are supposed to set `x[i, 0]` equal to `x[i, 0] / U[i, i]`, but since `U[i, i]`, must be one, the value of `x[i, 0]` won't change and we don't even need to call `gsl_matrix_uchar_set()`.

Reply:
I think 2. is the right way regarding what we should do to `x`. 

To explain why I did that change: I looked at the if/elseifs and noticed that it always sets the new value to the old value, unless the the old value `x_i` is 0 and `U_i_i` is 1, in which case it just errors.

So, you're right, we don't ever need to set things; this could handle `x` as an immutable vector. :exploding_head:

Reply:
> So, you're right, we don't ever need to set things; this could handle x as an immutable vector.

Well no, because somewhere else in this code there is the following operation:
```
// x[i, 0] = x[i, 0] + U[i, j] * x[j, 0]
```
We are doing Gauss reduction in place, after all, so `x` needs to be mutated in some way.

The main point of this comment was that division in GF(2) is a no-op, unless you are trying to divide by 0. And you do need to check for division by 0, because that will eventually happen if you try to reduce a matrix that is not invertible (which can happen if a bad alist is given).